book-content/chapters/16-the-utils-folder.md

```diff
@@ -0,0 +1,1527 @@
+
+```ts twoslash
+const removeId = <TObj extends { id: unknown }>(obj: TObj) => {
+  const { id, ...rest } = obj;
+  return rest;
+};
+// ---cut---
+const result = removeId({ id: 1, name: "Alice" });
+//    ^?
+```
+
+Note how clever TypeScript is being here. Even though we didn't specify a return type for `removeId`, TypeScript has inferred that `result` is an object with all the properties of the input object, except `id`.
+
+## Type Predicates
+
+We were introduced to type predicates way back in chapter 5, when we looked at narrowing. They're used to capture reusable logic that narrows the type of a variable.
+
+For example, say we want to ensure that a variable is an `Album` before we try accessing its properties or passing it to a function that requires an `Album`.
+
+We can write an `isAlbum` function that takes in an input, and checks for all the required properties.
+
+```typescript
+function isAlbum(input: unknown) {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input
+  );
+}
+```
+
+If we hover over `isAlbum`, we can see a rather ugly type signature:
+
+```typescript
+// hovering over isAlbum shows:
+function isAlbum(
+  input: unknown,
+): input is object &
+  Record<"id", unknown> &
+  Record<"title", unknown> &
+  Record<"artist", unknown> &
+  Record<"year", unknown>;
+```
+
+This is technically correct: a big intersection between an `object` and a bunch of `Record`s. But it's not very helpful.
+
+When we try to use `isAlbum` to narrow the type of a value, TypeScript won't infer it correctly:
+
+```ts twoslash
+// @errors: 18046
+function isAlbum(input: unknown) {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input
+  );
+}
+
+// ---cut---
+const run = (maybeAlbum: unknown) => {
+  if (isAlbum(maybeAlbum)) {
+    maybeAlbum.name.toUpperCase();
+  }
+};
+```
+
+To fix this, we'd need to add even more checks to `isAlbum` to ensure we're checking the types of all the properties:
+
+```typescript
+function isAlbum(input: unknown) {
+  return (
+    typeof input === "object" &&
+    input !== null &&
+    "id" in input &&
+    "title" in input &&
+    "artist" in input &&
+    "year" in input &&
+    typeof input.id === "number" &&
+    typeof input.title === "string" &&
+    typeof input.artist === "string" &&
+    typeof input.year === "number"
+  );
+}
+```
+
+But at this point, something frustrating happens - TypeScript _stops_ inferring the return value of the function. We can see this by hovering over `isAlbum`:
+
+```typescript
+// hovering over isAlbum shows:
+function isAlbum(input: unknown): boolean;
+```
+
+This is because TypeScript's type predicate inference has limits - it can only process a certain level of complexity.
```

Comment:
This is related to a pretty basic TS limitation that predates type predicate inference by a lot. Property refinements don't flow back into the "parent value":
```ts
function test(obj: { foo?: string | undefined }) {
  if (typeof obj.foo === "string") {
    const str: string = obj.foo; // ok
    const objWithStrFoo: { foo: string } = obj; // not ok 😭
  }
}
```

Specifically, with the given example, TS isn't currently smart enough to assign `input`'s type based on those `typeof input.id === 'number'` checks. And since it isn't capable of that... it doesn't quite know how to serialize this type for it to appear in the type predicate's annotation.

But even more importantly, we must understand that type predicates have to be exhaustive~. TS has to prove that if a type predicate receives some value then - based on those checks - it will always return true or false. And since the "final type" here is `input is object & Record<"id", unknown> & Record<"title", unknown> & Record<"artist", unknown> & Record<"year", unknown>;` it could actually return either true or false for a value matching this type (since the `typeof` property checks could yield either true or false)